server/saved_objects/saved_objects_wrapper.ts

```diff
@@ -0,0 +1,190 @@
+  SavedObjectsBulkUpdateOptions,
+  SavedObjectsBulkUpdateResponse,
+  SavedObjectsCheckConflictsObject,
+  SavedObjectsCheckConflictsResponse,
+  SavedObjectsClientWrapperFactory,
+  SavedObjectsCreateOptions,
+  SavedObjectsDeleteOptions,
+  SavedObjectsFindOptions,
+  SavedObjectsFindResponse,
+  SavedObjectsUpdateOptions,
+  SavedObjectsUpdateResponse,
+} from 'opensearch-dashboards/server';
+import { Config } from 'packages/osd-config/target';
+import { SecurityPluginConfigType } from '..';
+import {
+  DEFAULT_TENANT,
+  GLOBAL_TENANT,
+  isPrivateTenant,
+  PRIVATE_TENANT,
+} from '../../public/apps/configuration/utils/tenant-utils';
+import { OpenSearchDashboardsAuthState } from '../auth/types/authentication_type';
+
+export class SecuritySavedObjectsClientWrapper {
+  public httpStart?: HttpServiceStart;
+  public config?: SecurityPluginConfigType;
+
+  constructor() {}
+
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const state: OpenSearchDashboardsAuthState =
+      (this.httpStart!.auth.get(wrapperOptions.request).state as OpenSearchDashboardsAuthState) ||
+      {};
+
+    const selectedTenant = state.selectedTenant;
+    const username = state.authInfo?.user_name;
+    const isGlobalEnabled = this.config!.multitenancy.tenants.enable_global;
+    const isPrivateEnabled = this.config!.multitenancy.tenants.enable_private;
+
+    let namespaceValue = selectedTenant;
+
+    const createWithNamespace = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      if (selectedTenant !== undefined && isPrivateEnabled && isPrivateTenant(selectedTenant)) {
+        namespaceValue = selectedTenant + username;
+      }
+      _.assign(options, { namespace: [namespaceValue] });
+      return await wrapperOptions.client.create(type, attributes, options);
+    };
+
+    const bulkGetWithNamespace = async <T = unknown>(
+      objects: SavedObjectsBulkGetObject[] = [],
+      options: SavedObjectsBaseOptions = {}
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      if (selectedTenant !== undefined && isPrivateEnabled && isPrivateTenant(selectedTenant)) {
+        namespaceValue = selectedTenant + username;
+      }
+      _.assign(options, { namespace: [namespaceValue] });
+      return await wrapperOptions.client.bulkGet(objects, options);
+    };
+
+    const findWithNamespace = async <T = unknown>(
+      options: SavedObjectsFindOptions
+    ): Promise<SavedObjectsFindResponse<T>> => {
+      const tenants = state.authInfo?.tenants;
+      const availableTenantNames = Object.keys(tenants!);
+      availableTenantNames.push(DEFAULT_TENANT);
+      if (isGlobalEnabled) {
+        availableTenantNames.push(GLOBAL_TENANT);
+      }
+      if (isPrivateEnabled) {
+        availableTenantNames.push(PRIVATE_TENANT + state.authInfo?.user_name);
+      }
+      _.assign(options, { namespaces: availableTenantNames });
```

Comment:
@cliu123 I found that the `find` API supports a `typeToNamespacesMap` that can be used to limit the namespaces the API returns for a given object type. This is useful for the problem with `Advanced Settings` being given the same `id` across tenants. The map can be used in the following way:

```suggestion
      const typeToNamespacesMap = {};
      if (selectedTenant === '__user__') {
        namespaceValue = selectedTenant + username;
      }
      const searchTypes = Array.isArray(options.type) ? options.type : [options.type];
      searchTypes.forEach((t) => {
        if (t === 'config') {
          if ('namespaces' in options) {
            if (options.namespaces.includes(namespaceValue)) {
              typeToNamespacesMap[t] = [namespaceValue];
            }
          } else {
            typeToNamespacesMap[t] = [namespaceValue];
          }
        } else {
          if ('namespaces' in options) {
            typeToNamespacesMap[t] = options.namespaces;
          } else {
            typeToNamespacesMap[t] = availableTenantNames;
          }
        }
      });
      options.typeToNamespacesMap = new Map(Object.entries(typeToNamespacesMap));
      options.type = '';
      options.namespaces = [];
```